added validation for CFD CSCwp64296

psureshb opened the pull request.

psureshb committed:
Added rogue ep/coop exception mac check for the CFD CSCwp64296

psureshb committed:
rogue exception macs detection failure after stateless reload test method added with pytest files.

psureshb committed:
Revert "Added rogue ep/coop exception mac check for the CFD CSCwp64296"

This reverts commit 3fc5ebfd461559ac36300591bd3bca3e4c9187bc.

psureshb committed:
added test for rogue exception MACs missing due to missing presListeners of exceptcont tree

psureshb committed:
added the defect CSCwp64296 documentation.

psureshb committed:
recommended action edited, cdets link replaced by cloudapps link

psureshb committed:
added log files

psureshb committed:
removed the log files in the directory

psureshb committed:
Merge branch 'master' of github.com:psureshb/ACI-Pre-Upgrade-Validation-Script into psureshb

psureshb committed:
resolved merge conficts due to new method "apic_downgrade_compat_warning_check" merged in master

diwakarmishra1190 reviewed (approved):
LGTM 
tested Locally

psureshb committed:
refactor: Update rogue_ep_coop_exception_mac_check logic and test cases for version compatibility

psureshb committed:
doc update

diwakarmishra1190 reviewed (approved):
New changes are working fine tested locally.

takishida committed:
Merge branch 'master' into psureshb

takishida committed:
doc: Update the doc with proper conditions to check

takishida commented on aci-preupgrade-validation-script.py:
```diff
@@ -6012,6 +6012,48 @@ def apic_vmm_inventory_sync_faults_check(**kwargs):
+    # if cversion.newer_than("3.1(2v)") and tversion.older_than("6.1(3g)"):
+    if (
+    (cversion.same_as("5.2(3e)") or cversion.newer_than("5.2(3e)")) and
+    (cversion.same_as("6.0(3g)") or cversion.older_than("6.0(3g)")) and
```
Why 6.0(3g)? The enhancement that caused CSCwp64296 was introduced in the first release of 6.0(3) which is 6.0(3d).

psureshb replied:
version changed to 6.0(3d)

takishida commented on aci-preupgrade-validation-script.py:
```diff
@@ -6012,6 +6012,48 @@ def apic_vmm_inventory_sync_faults_check(**kwargs):
+    )
+    ):
+        # endpoint to fetch the rogue exception MACs
+        exception_mac_api = 'fvRogueExceptionMac.json?query-target-filter=and(wcard(fvRogueExceptionMac.dn,"([0-9a-fA-F]{2}:){5}[0-9a-fA-F]{2}"))'
```
Why are you matching MAC address in regex here? `fvRogueExceptionMac` is for MAC as the name suggests. Its DN always has a MAC address.

Unless I'm missing something, this is just causing unnecessary query work on the APICs.

psureshb replied:
removed the logic of matching MAC address using regex

takishida commented on aci-preupgrade-validation-script.py:
```diff
@@ -6012,6 +6012,48 @@ def apic_vmm_inventory_sync_faults_check(**kwargs):
+        exception_mac_api = 'fvRogueExceptionMac.json?query-target-filter=and(wcard(fvRogueExceptionMac.dn,"([0-9a-fA-F]{2}:){5}[0-9a-fA-F]{2}"))'
+
+        # endpoint to fetch the presListener entries
+        presListener_api = 'presListener.json?query-target-filter=and(wcard(presListener.dn,"exceptcont"))'
```
Instead of regex(`wcard`)), you should do exact match with `lstDn` such as `eq(presListener.lstDn,"exceptcont")`

And use `rsp-subtree-include=count` instead of having APICs return all of the contents because what we need to know is only the number of objects.

psureshb replied:
using `eq(presListener.lstDn,"exceptcont")`  and `rsp-subtree-include=count`

takishida commented on aci-preupgrade-validation-script.py:
```diff
@@ -6012,6 +6012,48 @@ def apic_vmm_inventory_sync_faults_check(**kwargs):
+        exception_macs = icurl('class', exception_mac_api)
+
+        if exception_macs:
+            prints("Found {} exception MACs, checking presListener entries...".format(len(exception_macs)))
```
Do not use `prints()` inside a check function.

In the new framework with the progress bar, it will end up like this, breaking the progress bar view. Please use logging instead, if needed. Otherwise, those info should be stored in the `data` so that it can be displayed in the result table and JSON file.

```
--- skip ---
Collecting VPC Node IDs...

Found 1 exception MACs, checking presListener entries...-------------------------------------------------------| 0/1 checks completed     <--- !!!!!
Progress: |████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 checks completed


=== Check Result (failed only) ===
```

psureshb replied:
removed prints staments

takishida commented on aci-preupgrade-validation-script.py:
```diff
@@ -6012,6 +6012,48 @@ def apic_vmm_inventory_sync_faults_check(**kwargs):
+
+    # Target version check
+    if not tversion:
+        prints("Target version not provided, skipping check.")
```
```suggestion
```

takishida reviewed (changes requested):
@psureshb  @diwakarmishra1190 

I've updated the doc with the information that is relevant to users along with the new conditions.

Correct me if I'm wrong, but I doubt that the current condition checked by this rule will ever hit because I believe `presListener` will be missing on APICs after APICs are upgraded. But you are checking the current version being the older one and the target version being the affected one. That is only the case BEFORE APICs (or anything) are upgraded. In other words, the `presListener` issue has yet to happen.

Or if `presLIstener`s are created on APICs when switches are upgraded instead of APICs, using that as a condition doesn't make sense at all. This script is to prevent the issue from happening by detecting the conditions before the issue happens instead of checking the symptoms of the issue after hitting it.

Can you please review the updated doc and update the rule logic accordingly?

psureshb committed:
refactor: Update rogue_ep_coop_exception_mac_check logic and test cases for presListener API changes

- Adjusted the API queries in the rogue_ep_coop_exception_mac_check function to improve accuracy.
- Updated test cases to reflect changes in the expected structure of presListener responses.
- Added new test cases for scenarios with varying counts of presListener entries.
- Removed outdated comments and improved logging for better clarity.

psureshb committed:
refactor: Enhance rogue_ep_coop_exception_mac_check for improved validation of exception MACs and presListener entries

- Updated API queries to check for rogue exception MACs and presListener entries before and after APIC upgrades.
- Improved logging to provide clearer insights on the counts of exception MACs and presListener entries.
- Adjusted logic to handle scenarios with varying counts of presListener entries more effectively.

psureshb committed:
Merge branch 'psureshb' of github.com:psureshb/ACI-Pre-Upgrade-Validation-Script into psureshb

psureshb committed:
refactor: Revise rogue_ep_coop_exception_mac_check and documentation for clarity and accuracy

- Replaced the existing rogue_ep_coop_exception_mac_check function with a refined version that improves validation logic for rogue exception MACs and presListener entries.
- Updated documentation to reflect the new check title and conditions for the rogue EP Exception List.
- Enhanced test cases to align with the updated API response structure and added scenarios for post-APIC upgrade checks.
- Adjusted JSON test responses to ensure accurate validation during testing.

psureshb committed:
fix: Update exception MAC API query and test cases for version accuracy

- Modified the exception_mac_api query to include response subtree count for better validation.
- Corrected version numbers in test cases to reflect accurate edge and target versions.
- Added new test scenarios for post-APIC upgrade checks to enhance coverage and validation accuracy.

psureshb commented:
[pytest logs.txt](https://github.com/user-attachments/files/25971211/pytest.logs.txt)
[preupgrade_validator_logs.zip](https://github.com/user-attachments/files/25971306/preupgrade_validator_logs.zip)

lovkeshsharma702 commented on docs/docs/validations.md:
```diff
@@ -2668,6 +2672,33 @@ To avoid this risk, consider disabling Auto Firmware Update before upgrading to
 
+### Rogue EP Exception List missing on switches
+
+Rogue Endpoint Control and COOP Dampening are features that mitigate the impact of flapping endpoints by temporarily pausing the learning of such endpoints. However, in some environments, certain MAC or IP addresses are expected to move frequently.
```
remove 2677, no need to educate Cu on basic stuff. also remove all extra stuff.  just keep info related to bug. 

"The **Rogue/COOP Exception List** feature, introduced in 5.2(3), allows you to exclude specific MAC addresses from Rogue Endpoint Control and COOP Dampening. Initially, each MAC address had to be configured individually in each bridge domain. In 6.0(3), this feature was enhanced to support fabric-wide exception lists with wildcard options per bridge domain and the ability to exclude MAC addresses in L3Outs. 

due to [CSCwp64296][64], when upgrading spine switches to version 6.0(3)+ from an older version with **Rogue/COOP Exception List**s configured, some exception lists config may not be pushed to the spine switches after reboot. As a result, the feature may stop functioning after the upgrade. 

    The root cause is that internal objects called `presListener` for **Rogue/COOP Exception List**, which publish the configuration from APICs to switches, may be missing on the APICs after an upgrade. This is due to [CSCwp64296][64], introduced with the enhancement in 6.0(3).

Recommended action: delete the affected exception list and create it again. If needed, contact Cisco TAC to help recover missing `presListener` objects on APICs. 
" Keep only relevent info. No extra symbol or AI marking.

psureshb replied:
removed line 2677
Removed the additional data.

lovkeshsharma702 commented on tests/checks/rogue_ep_coop_exception_mac_check/test_rogue_ep_coop_exception_mac_check.py:
```diff
@@ -0,0 +1,212 @@
+        ),
+    ],
+    ids=[
+        "PASS_non_affected_tversion_6.1(4h)_fixed_affected_cversion_no_exception_MACs",
```
follow the structure used in old bug.

psureshb replied:
removed the ids for compliance with the standard structure used in other pytest files.

lovkeshsharma702 reviewed (commented):
work on comments.

psureshb committed:
refactor: Update documentation and test cases for Rogue/COOP Exception List to enhance clarity and accuracy
Updated pytest file and structured to comply with old cases

lovkeshsharma702 commented on docs/docs/validations.md:
```diff
@@ -2668,6 +2672,17 @@ To avoid this risk, consider disabling Auto Firmware Update before upgrading to
+
+However, due to [CSCwp64296][64], when upgrading spine switches to version 6.0(3)+ from an older version with Rogue/COOP Exception Lists configured, some exception lists may not be pushed to the spine switches. As a result, the feature may stop functioning after the upgrade. 
+
+The root cause is that internal objects called `presListener` for Rogue/COOP Exception List, which publish the configuration from APICs to switches, may be missing on the APICs after an upgrade. This is due to [CSCwp64296][64], introduced with the enhancement in 6.0(3).
```
we do not talk about the regression " This is due to [CSCwp64296][64], introduced with the enhancement in 6.0(3)." remove this.

psureshb committed:
fix: Clarify root cause of missing presListener objects for Rogue/COOP Exception List after upgrade. Removed the regression context from root cause statement

takishida committed:
fix: Add proper version handling with NA

takishida committed:
chore: More verbose output. pytest with expected_data

takishida commented:
## Example result outputs:

Pre-APIC Upgrade
```
[Check 85/91] Rogue EP Exception List missing on switches...                                                       FAIL - OUTAGE WARNING!!
  Rogue Exception MACs Count  presListener Count
  --------------------------  ------------------
  5                           N/A

  Recommended Action: Change the target version to a fixed version of CSCwp64296.
  Reference Document: https://datacenter.github.io/ACI-Pre-Upgrade-Validation-Script/validations/#rogue-ep-exception-list-missing-on-switches
```


Post-APIC Upgrade
```
[Check 85/91] Rogue EP Exception List missing on switches...                                                       FAIL - OUTAGE WARNING!!
  Rogue Exception MACs Count  presListener Count
  --------------------------  ------------------
  5                           only 27 found out of 32

  Recommended Action: Delete the exception lists and create again before upgrading switches. Or contact Cisco TAC to restore the missing presListener objects.
  Reference Document: https://datacenter.github.io/ACI-Pre-Upgrade-Validation-Script/validations/#rogue-ep-exception-list-missing-on-switches
```
